### PR TITLE
napi: add finalize hint support

### PR DESCRIPTION
**Pull request by digitalinfinity**
Change napi_finalize_callback to support a hint parameter. APIs that take in a finalize callback now also take a hint. This hint is passed back in when the finalize callback is called

Fixes: #111

### CODE REVIEW

**Comment by digitalinfinity** [comment]
@mhdawson @gabrielschulhof @sampsongao feel free to review too if you guys have time

/cc @kkoopa

**Comment by digitalinfinity** [comment]
CI in progress at https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-commit-linux-abi/38/

File: test/addons-abi/test_buffer/test_buffer.cc [hunk 1]
```diff
@@ -52,7 +52,12 @@ void newExternalBuffer(napi_env env, napi_callback_info info) {
   JS_ASSERT(env, theCopy, "Failed to copy static text for newExternalBuffer");
   NAPI_CALL(env,
             napi_create_external_buffer(
-                env, sizeof(theText), theCopy, deleteTheText, &theBuffer));
+                env,
+                sizeof(theText),
+                theCopy,
+                deleteTheText,  // finalize_hint
```
Comment (jasongin) [inline test/addons-abi/test_buffer/test_buffer.cc]: The comment is on the wrong line: the `nullptr` argument is the hint.

**Review by jasongin — APPROVED** [review]
LGTM

File: test/addons-abi/test_buffer/test_buffer.cc (hunk 1 above)
Comment (digitalinfinity) [reply test/addons-abi/test_buffer/test_buffer.cc]: Whoops! Thanks, pushed a commit to fix it

**Review by boingoing — APPROVED** [review]
LGTM

**Review by mhdawson — APPROVED** [review]
LGTM

**Comment by digitalinfinity** [comment]
Thanks everyone for the reviews- merging now!

**Pull request merged**